grandpa/src/consensus.rs

```diff
@@ -0,0 +1,295 @@
+            )))
+        }
+
+        let first_headers = AncestryChain::<H>::new(&first_proof.unknown_headers);
+        let first_target =
+            first_proof.unknown_headers.iter().max_by_key(|h| *h.number()).ok_or_else(|| {
+                Error::ImplementationSpecific(format!("Unknown headers can't be empty!"))
+            })?;
+
+        let second_headers = AncestryChain::<H>::new(&second_proof.unknown_headers);
+        let second_target =
+            second_proof.unknown_headers.iter().max_by_key(|h| *h.number()).ok_or_else(|| {
+                Error::ImplementationSpecific(format!("Unknown headers can't be empty!"))
+            })?;
+
+        if first_target.hash() != first_proof.block || second_target.hash() != second_proof.block {
+            return Err(Error::ImplementationSpecific(format!(
+                "Fraud proofs are not for the same chain"
+            )))
+        }
+
+        let first_base =
+            first_proof.unknown_headers.iter().min_by_key(|h| *h.number()).ok_or_else(|| {
+                Error::ImplementationSpecific(format!("Unknown headers can't be empty!"))
+            })?;
+        first_headers
+            .ancestry(first_base.hash(), first_target.hash())
+            .map_err(|_| Error::ImplementationSpecific(format!("Invalid ancestry!")))?;
+
+        let second_base =
+            second_proof.unknown_headers.iter().min_by_key(|h| *h.number()).ok_or_else(|| {
+                Error::ImplementationSpecific(format!("Unknown headers can't be empty!"))
+            })?;
+        second_headers
+            .ancestry(second_base.hash(), second_target.hash())
+            .map_err(|_| Error::ImplementationSpecific(format!("Invalid ancestry!")))?;
+
+        let first_parent = first_base.parent_hash();
+        let second_parent = second_base.parent_hash();
+
+        if first_parent != second_parent {
+            return Err(Error::ImplementationSpecific(format!(
+                "Fraud proofs are not for the same ancestor"
+            )))
+        }
+
+        //TODO: how to verify that first parent contains relay header hash
+
+        let first_justification =
+            GrandpaJustification::<H>::decode(&mut &first_proof.justification[..]).map_err(
+                |_| Error::ImplementationSpecific(format!("Could not decode first justification")),
+            )?;
+
+        let second_justification =
+            GrandpaJustification::<H>::decode(&mut &second_proof.justification[..]).map_err(
+                |_| Error::ImplementationSpecific(format!("Could not decode second justification")),
+            )?;
+
+        if first_proof.block != first_justification.commit.target_hash ||
+            second_proof.block != second_justification.commit.target_hash
+        {
+            Err(Error::ImplementationSpecific(
+                format!("First or second finality proof block hash does not match justification target hash")
+            ))?
+        }
+
```

Comment:
Check that the target_hash for either the first or second justification is equal to the latest_hash on the consensus state.